include/deal.II/base/memory_space.h

```diff
@@ -0,0 +1,212 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii_memory_space_h
+#define dealii_memory_space_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/cuda.h>
+#include <deal.II/base/exceptions.h>
+
+#include <memory>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ *
+ */
+namespace MemorySpace
+{
+  /**
+   * Structure describing Host memory space.
+   */
+  struct Host
+  {};
+
+
+
+  /**
+   * Structure describing CUDA memory space.
+   */
+  struct CUDA
+  {};
+
+
+
+#if defined(DEAL_II_WITH_CUDA) && defined(__CUDACC__)
+  /**
+   * Deleter for unique/shared pointer pointing to an array allocated on the
+   * device.
+   */
+  template <typename Number>
+  void
+  deleter_device_array(Number *ptr) noexcept
+  {
+    const cudaError_t cuda_error_code = cudaFree(ptr);
+    (void)cuda_error_code;
+    AssertNothrow(cuda_error_code == cudaSuccess,
+                  ExcCudaError(cudaGetErrorString(cuda_error_code)));
+  }
+#endif
```

Comment:
We might want to put this function into a more central place like `base/cuda.h` at some point. We now have this also for `CUDAWrappers::Vector` and `CUDAWrappers::Precondition`. We can, of course, do that in later PRs.

Reply:
Yes let's do that in another PR once this has been merged.